src/java.base/share/classes/java/lang/reflect/code/writer/OpWriter.java

```diff
@@ -48,16 +48,20 @@ public final class OpWriter {
     static final class GlobalValueBlockNaming implements Function<CodeItem, String> {
         final Map<CodeItem, String> gn;
         int valueOrdinal = 0;
-        int blockOrdinal = 0;
 
         GlobalValueBlockNaming() {
             this.gn = new HashMap<>();
         }
 
+        private String name(Block b) {
+            Block p = b.parentBody().parentOp().parentBlock();
+            return (p == null ? "block_" : name(p) + "_") + b.index();
+        }
+
         @Override
         public String apply(CodeItem codeItem) {
             return switch (codeItem) {
-                case Block block -> gn.computeIfAbsent(block, _b -> "block_" + blockOrdinal++);
+                case Block block -> gn.computeIfAbsent(block, _b -> name(block));
```

Comment:
```suggestion
                case Block block -> gn.computeIfAbsent(block, _b -> name(_b));
```
Or replace with method reference to static method?

Reply:
unfortunately `_b` is not a `Block` but `CodeItem`